### PR TITLE
ntp configurable

### PR DESCRIPTION
**Pull request by abashir236**
This PR:

- adds/changes/removes etc

### Testing

Description on how {APP-NAME} can be tested.

- [ ] fresh install works
- [ ] upgrade from previous version works

#### Other testing

Description of features to additionally test for {APP-NAME} installations.

- [ ] check reconciliation of existing resources after upgrading
- [ ] X still works after upgrade
- [ ] Y is installed correctly

### Checklist

- [ ] Update changelog in CHANGELOG.md.
- [ ] Make sure `values.yaml` and `values.schema.json` are valid.
- [ ] Update Lastpass values if required.

### CODE REVIEW

File: helm/cluster-cloud-director/templates/coredns.yaml [hunk 1]
```diff
@@ -1,2 +1,2 @@
 # use the `_coredns.tpl` from `cluster-shared`
-{{- include "coredns" .}}
+{{- include "shared.coredns-adopter" .}}
```
Comment (vxav) [inline helm/cluster-cloud-director/templates/coredns.yaml:2]: Why did you change this?

File: helm/cluster-cloud-director/templates/_helpers.tpl [hunk 2]
```diff
@@ -130,6 +130,12 @@ files:
 {{- if $.Values.network.staticRoutes }}
 {{- include "staticRoutes" . | nindent 2}}
 {{- end }}
+ntp:
+{{- if $.Values.ntp.enabled }}
+servers:
+{{- range .servers }}
+ - {{ . }}
+{{- end }}
```
Comment (vxav) [inline helm/cluster-cloud-director/templates/_helpers.tpl:138]: You are missing a `{{- end }}` for the `if` statement.

File: helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml [hunk 3]
```diff
@@ -111,6 +111,12 @@ spec:
       {{- if $.Values.network.staticRoutes }}
       {{- include "staticRoutes" . | nindent 6}}
       {{- end }}
+    ntp:
+      {{- if .Values.ntp.enabled }}
+      servers:
+      {{- range .Values.ntp.servers }}
+      - {{ . }}
+      {{- end }}
```
Comment (vxav) [inline helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml:119]: Here as well you are missing a `{{- end }}` for the `if` statement.

File: helm/cluster-cloud-director/templates/_helpers.tpl [hunk 4]
```diff
@@ -130,6 +130,12 @@ files:
 {{- if $.Values.network.staticRoutes }}
 {{- include "staticRoutes" . | nindent 2}}
 {{- end }}
+ntp:
+{{- if $.Values.ntp.enabled }}
```
Comment (erkanerol) [inline helm/cluster-cloud-director/templates/_helpers.tpl:134]: Please move this if before `ntp:`

File: helm/cluster-cloud-director/templates/_helpers.tpl [hunk 5]
```diff
@@ -130,6 +130,12 @@ files:
...
+servers:
```
Comment (erkanerol) [inline helm/cluster-cloud-director/templates/_helpers.tpl:135]: missing indentation?

File: helm/cluster-cloud-director/templates/_helpers.tpl [hunk 6]
```diff
@@ -141,7 +147,7 @@ postKubeadmCommands:
 - systemctl enable --now static-routes.service
 {{- end }}
 {{- end -}}
-
+{{- end -}}
```
Comment (erkanerol) [inline helm/cluster-cloud-director/templates/_helpers.tpl:150]: This should be before `preKubeadmCommands`

File: helm/cluster-cloud-director/values.yaml [hunk 7]
```diff
@@ -92,6 +92,10 @@ nodePools: {}
   # worker:
   #   class: default  # Node classes name (defined above)
   #   replicas: 2
+ntp:
+  servers: 
+  - ""
+  enabled: false
```
Comment (vxav) [inline helm/cluster-cloud-director/values.yaml:98]: Maybe we don't need the `enabled` switch and only add the ntp config to the chart if `servers` is populated.

File: helm/cluster-cloud-director/values.yaml [hunk 8]
```diff
@@ -92,6 +92,10 @@ nodePools: {}
   # worker:
   #   class: default  # Node classes name (defined above)
   #   replicas: 2
+ntp:
+  servers:
```
Comment (erkanerol) [inline helm/cluster-cloud-director/values.yaml:96]: default value should be `servers: []`

File: helm/cluster-cloud-director/values.schema.json [hunk 9]
```diff
@@ -269,6 +269,17 @@
                         }
                     }
                 },
+                "ntp": {
+                    "type": "object",
+                    "properties": {
+                        "server": {
+                            "type": "string"
```
Comment (erkanerol) [inline helm/cluster-cloud-director/values.schema.json:276]: isn't it string array instead of string?

File: helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml [hunk 10]
```diff
@@ -131,3 +137,4 @@ spec:
       namespace: {{ .Release.Namespace }}
   replicas: {{ .Values.controlPlane.replicas }}
   version: {{ .Values.kubernetesVersion }}
+  {{- end }}
```
Comment (erkanerol) [inline helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml:140]: new line

File: helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml [hunk 11]
```diff
@@ -111,6 +111,12 @@ spec:
       {{- if $.Values.network.staticRoutes }}
       {{- include "staticRoutes" . | nindent 6}}
       {{- end }}
+    ntp:
+      {{- if .Values.ntp.enabled }}
```
Comment (erkanerol) [inline helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml:115]: move this part to a common function and call it in kubeadmcontrolplane and kubeadmconfig spec.

File: helm/cluster-cloud-director/templates/coredns.yaml (hunk 1 above)
Comment (erkanerol) [inline helm/cluster-cloud-director/templates/coredns.yaml:2]: are you sure about this?

**Review by erkanerol — COMMENTED** [review]
Please remove `.DS_Store` files from the PR.

File: CHANGELOG.md [hunk 12]
```diff
@@ -22,6 +22,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Add configurable ntp on nodes.
```
Comment (vxav) [inline CHANGELOG.md:25]: This should be under `[Unreleased]`

File: helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml [hunk 13]
```diff
@@ -109,6 +109,7 @@ spec:
       {{- if $.Values.network.staticRoutes }}
       {{- include "staticRoutes" . | nindent 6}}
       {{- end }}
+      {{- include "ntpServers" . | nindent 4}}
```
Comment (vxav) [inline helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml:112]: @erkanerol This is correct obviously but I find it a bit confusing as the `ntp:` section name is defined in the template, here it looks like it is under `files:`. 

As a personal preference, I'd rather see the `ntp:` section explicitly here. But it's only cosmetic.

**Review by kopiczko — COMMENTED** [review]
LGTM when changelog is fixed

File: helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml (hunk 13 above)
Comment (erkanerol) [reply helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml:112]: But then we need to add `if-else` logic here, which makes this file more complex. I prefer encapsulating features with helper functions.

**Comment by erkanerol** [comment]
We (me+Zak) collaborated on this PR and tested it in grenoble. It was not effective on the nodes when we checked it.  @abashir236  took responsibility for ensuring the configuration of the nodes. Until getting OK from him, please don't merge this PR.

File: helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml (hunk 13 above)
Comment (jkremser) [reply helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml:112]: moving the whole `{{- include "ntpServers" . | nindent 4}}` 2 spaces to the left should definitely make it more readable. Now it's under files section, but correctly indented by 4 spaces

Comment (erkanerol) [reply helm/cluster-cloud-director/templates/kubeadmcontrolplane.yaml]: Totally agree and fixed.

**Comment by marians** [comment]
Headsup: `values.schema.json` has just changed via https://github.com/giantswarm/cluster-cloud-director/pull/91

To enable merging your PR, please apply `schemalint normalize` to the `values.schema.json` file in your branch. After that, things should clear up nicely. Please ping me if you have any questions.

**Comment by vxav** [comment]
FYI: After testing this PR, I propose a different approach to make it work.
https://github.com/giantswarm/cluster-cloud-director/pull/101/

**Comment by erkanerol** [comment]
Do you know why this doesn't work?

**Comment by vxav** [comment]
After testing a bit, the kubeadm implementation doesn't seem to work. Only resources I could find was this [old issue](https://github.com/kubernetes-retired/cluster-api-bootstrap-provider-kubeadm/issues/101). Also the `chrony` config takes both `servers` and `pools` but kubeadm only takes `servers` so I think we should open this a bit.
